pkg/webhook/webhook.go

```diff
@@ -0,0 +1,201 @@
+	supportedVersions = map[string][]string{
+		"0.1": []string{},
+		"0.2": []string{},
+		"0.3": []string{},
+		"0.4": []string{},
+	}
+)
+
+// Create webhook server and register webhook to it
+func Add(mgr manager.Manager) error {
+	log.Printf("Initializing webhook")
+	svr, err := newWebhookServer(mgr)
+	if err != nil {
+		log.Printf("Failed to create webhook server: %v", err)
+		return err
+	}
+
+	wh, err := newValidatingWebhook(mgr)
+	if err != nil {
+		log.Printf("Failed to create validating webhook: %v", err)
+		return err
+	}
+
+	svr.Register(wh)
+	return nil
+}
+
+func newValidatingWebhook(mgr manager.Manager) (*admission.Webhook, error) {
+	return builder.NewWebhookBuilder().
+		Validating().
+		Operations(admissionregistrationv1beta1.Create).
+		ForType(&pravegav1alpha1.PravegaCluster{}).
+		Handlers(&pravegaWebhookHandler{}).
+		WithManager(mgr).
+		Build()
+}
+
+func newWebhookServer(mgr manager.Manager) (*webhook.Server, error) {
+	return webhook.NewServer("admission-webhook-server", mgr, webhook.ServerOptions{
+		CertDir: CertDir,
+		BootstrapOptions: &webhook.BootstrapOptions{
+			// TODO: garbage collect webhook k8s service
+			Service: &webhook.Service{
+				Namespace: os.Getenv("WATCH_NAMESPACE"),
+				Name:      "admission-webhook-server-service",
+				Selectors: map[string]string{
+					"component": "pravega-operator",
+				},
+			},
+		},
+	})
+}
+
+type pravegaWebhookHandler struct {
+	client  client.Client
+	scheme  *runtime.Scheme
+	decoder admissiontypes.Decoder
+}
+
+var _ admission.Handler = &pravegaWebhookHandler{}
+
+// Webhook server will call this func when request comes in
+func (pwh *pravegaWebhookHandler) Handle(ctx context.Context, req admissiontypes.Request) admissiontypes.Response {
+	log.Printf("Webhook is handling incoming requests")
+	pravega := &pravegav1alpha1.PravegaCluster{}
+
+	if err := pwh.decoder.Decode(req, pravega); err != nil {
+		return admission.ErrorResponse(http.StatusBadRequest, err)
+	}
+	copy := pravega.DeepCopy()
+
+	code, err := pwh.validatePravegaManifest(ctx, copy)
+
+	switch code {
+	case "400":
+		return admission.ErrorResponse(http.StatusBadRequest, err)
+	case "500":
+		return admission.ErrorResponse(http.StatusInternalServerError, err)
+	}
+
+	return admission.ValidationResponse(true, "")
+}
+
+func (pwh *pravegaWebhookHandler) validatePravegaManifest(ctx context.Context, p *pravegav1alpha1.PravegaCluster) (string, error) {
+	if code, error := pwh.validatePravegaVersion(ctx, p); code != "200" {
+		return code, error
+	}
+
+	//TODO: Add other validators here
+	return "200", nil
+}
+
+func (pwh *pravegaWebhookHandler) validatePravegaVersion(ctx context.Context, p *pravegav1alpha1.PravegaCluster) (string, error) {
+	// Check if the request has a legal Pravega version
+	if p.Spec.Version == "" {
+		if p.Spec.Pravega != nil && p.Spec.Pravega.Image != nil && p.Spec.Pravega.Image.Tag != "" {
+			p.Spec.Version = p.Spec.Pravega.Image.Tag
```

Comment:
According to the [documentation](https://kubernetes.io/docs/reference/access-authn-authz/admission-controllers/#validatingadmissionwebhook) `ValidatingAdmissionWebhook` should not modify the object. `MutatingAdmissionWebhook` are used for that. 

`MutatingAdmissionWebhook` run before `ValidatingAdmissionWebhook`, so maybe we can create a `MutatingAdmissionWebhook` that just applies the default values to the object, i.e. `p.withDefaults()`. 

What do you think?

Reply:
Sorry I will make it more clear!